src/haz3lcore/statics/Statics.re

```diff
@@ -260,10 +273,34 @@ and uexp_to_info_map =
     );
   | Let(p, def, body) =>
     let (p_syn, _m) = go_pat(~is_synswitch=true, ~mode=Syn, p, m);
-    let def_ctx = extend_let_def_ctx(ctx, p, p_syn.ctx, def);
-    let (def, m) = go'(~ctx=def_ctx, ~mode=Ana(p_syn.ty), def, m);
-    /* Analyze pattern to incorporate def type into ctx */
-    let (p_ana, m) = go_pat(~is_synswitch=false, ~mode=Ana(def.ty), p, m);
+    let (def, p_ana, m) =
+      if (!is_recursive(ctx, p, def, p_syn.ty)) {
+        let (def, m) = go(~mode=Ana(p_syn.ty), def, m);
+        let (p_ana, m) =
+          go_pat(~is_synswitch=false, ~mode=Ana(def.ty), p, m);
+        (def, p_ana, m);
+      } else {
+        let (def_base, _m) = go'(~ctx=p_syn.ctx, ~mode=Ana(p_syn.ty), def, m) /* Analyze pattern to incorporate def type into ctx */;
+        let (p_ana, m) =
+          go_pat(~is_synswitch=false, ~mode=Ana(def_base.ty), p, m);
+        let def_ctx = p_ana.ctx;
+        let (def_base2, _m) =
```

Comment:
The following pieces of code aims to keep the synthesized types consistent.
For example, with the new context for the definition expression, `odd` has type `? -> Bool`, allowing `fun x -> odd(x - 1)` to synthesize out a more accurate type `? -> Bool`, which is consistent but not equal to the "finalized" type for `even`, `? -> ?`.
So if we can synthesize out a more accurate (than the "finalized" context) type for some variable(s), instead of doing so, we use the synthesized type to analyze against the expected type in the "finalized" context:
![image](https://github.com/hazelgrove/hazel/assets/108375845/2b4b61f6-2d97-4dc4-bbda-fffd58d2bc83)

I'm considering factoring out this piece of code but have no idea where to place it...